swift/apple/FirezoneKit/Sources/FirezoneKit/Stores/Store.swift

```diff
@@ -0,0 +1,196 @@
+
+  private func initNotifications() {
+    // Finish initializing notification binding
+    sessionNotification.signInHandler = {
+      Task { await WebAuthSession.signIn(store: self) }
+    }
+
+    sessionNotification.$decision
+      .receive(on: DispatchQueue.main)
+      .sink(receiveValue: { [weak self] decision in
+        guard let self = self else { return }
+        self.decision = decision
+      })
+      .store(in: &cancellables)
+  }
+
+  private func initTunnelManager() {
+    // Subscribe to status updates from the tunnelManager
+    tunnelManager.statusChangeHandler = handleVPNStatusChange
+
+    // Load our existing VPN profile and initialize our state
+    tunnelManager.load() { loadedStatus, loadedSettings, loadedActorName in
+      DispatchQueue.main.async {
+        self.status = loadedStatus
+
+        if let loadedSettings = loadedSettings {
+          self.settings = loadedSettings
+        }
+
+        if let loadedActorName = loadedActorName {
+          self.actorName = loadedActorName
+        }
+
+        // Try to connect on app launch
+        if self.status == .disconnected {
+          Task { try await self.start() }
+        }
+      }
+    }
+  }
+
+  func requestNotifications() {
+    #if os(iOS)
+    sessionNotification.askUserForNotificationPermissions()
+    #endif
+  }
+
+  func createVPNProfile() {
+    DispatchQueue.main.async {
+      Task {
+        self.settings = try await self.tunnelManager.create()
+      }
+    }
+  }
+
+  func authURL() -> URL? {
+    return URL(string: settings.authBaseURL)
+  }
+
+  func start(token: String? = nil) async throws {
+    guard status == .disconnected
+    else {
+      Log.app.log("\(#function): Already connected")
+      return
+    }
+
+    tunnelManager.start(token: token)
+  }
+
+  func stop(clearToken: Bool = false) {
+    guard [.connected, .connecting, .reasserting].contains(status)
+    else {
+      Log.app.error("\(#function): Can't stop from state: \(status)")
+      return
+    }
+
+    tunnelManager.stop(clearToken: clearToken)
+  }
+
+  func signIn(authResponse: AuthResponse) async throws {
+    // Save actorName
+    DispatchQueue.main.async { self.actorName = authResponse.actorName }
+
+    try await tunnelManager.saveSettings(settings)
+    try await tunnelManager.saveActorName(authResponse.actorName)
+
+    // Bring the tunnel up and send it a token to start
+    tunnelManager.start(token: authResponse.token)
+  }
+
+  func signOut() async throws {
+    // Stop tunnel and clear token
+    stop(clearToken: true)
+  }
+
+  // Network Extensions don't have a 2-way binding up to the GUI process,
+  // so we need to periodically ask the tunnel process for them.
+  func beginUpdatingResources(callback: @escaping (Data) -> Void) {
+    Log.app.log("\(#function)")
+
+    tunnelManager.fetchResources(callback: callback)
+    let intervalInSeconds: TimeInterval = 1
+    let timer = Timer(timeInterval: intervalInSeconds, repeats: true) { [weak self] _ in
```

Comment:
Oh, this runs forever. So this `beginUpdatingResources` is only called at app startup, right? Is that a change? I thought it used to be called when you open the menu bar.

Reply:
It's called whenever the tunnel moves into `.connected` status. Before, it was started whenever the menu was shown _and_ the tunnel moved into `.connected` status which meant observing two properties and preventing races between them (Publishers.combineLatest would probably have worked)

In the end I thought the added complexity is not worth it because the fetching of resources is basically a no-op if they haven't changed

Reply:
So it never runs two of these timers, it'll call `endUpdatingResources` when the tunnel disconnects, before the 2nd `beginUpdatingResources`?

Reply:
Actually, that's not quite accurate -- it doesn't start two timers because it's started and stopped based on the tunnel state entering `.connected` and leaving that state.

I suppose we could invalidate the timer at the start of this function "just in case", but hmm I was trying to get rid of defensive code that the client was littered with because we weren't confident in what was happening and just tried to handle every possible state things could be in.

I'm pretty confident this is tied reliably to the tunnel state, so we can stick with that and learn otherwise.

Reply:
Good points. I tend to agree. The timer not being nil should be tested in a unit test -- I don't think Swift has inline asserts.

The mess we were in prior was that we had the worst of both worlds, no tests to validate assumptions, and no assumptions to learn about how we might be wrong.

Reply:
ChatGPT says it does. I tried this in Godbolt
```swift
func panic_if_zero(_ x: Int32) -> Bool {
    assert(x != 0)
    return x == 42
}
```
https://godbolt.org/z/PW65srhrE

Unit tests are good but if we're outside the tested space in production we'd probably want something to happen

Reply:
Whoops.

Maybe `precondition` then. https://developer.apple.com/documentation/swift/precondition(_:_:file:line:)
Says it's checked even in "shipping" code.